pkg/test/util/structpath/instance.go

```diff
@@ -221,13 +226,23 @@ func (i *Instance) NotExists(path string, args ...interface{}) *Instance {
 		}
 		return nil
 	})
-	return i
 }
 
-// Check executes the set of contstraints for this selection
+// Check executes the set of constraints for this selection
 // and returns the first error encountered, or nil if all constraints
-// have been successfully met.
+// have been successfully met. All constraints are removed after them
+// check is performed.
 func (i *Instance) Check() error {
+	// After the check completes, clear out the constraints.
+	defer func() {
```

Comment:
Why do you need to do this?

Reply:
All of the constraints that have accumulated are targeted to the next Check call.  It seems a valid use case to call check on one group of expressions and then call check on another group and to then do a logical operationr (AND, OR) on the results of the two check calls.

This was in fact, something I'm doing in one of my other PRs.

Reply:
Wouldn't it be safer to make a copy of the instance and return it from the appendConstraint? That way the constraints are always add-only, and if the instances are somehow aliased (i.e. copy the reference to another variable), they will behave exactly the same way, regardless of how many times they are called.

Reply:
It could lead to unexpected oddities in usage.  For example:

v.Equals(ExpressionA)

v..Equals(ExpressionB)

v.Check()

Would not evaluate Expression A or ExpressionB.

I understand that you're trying to build immutability into it, but I'm not sure it makes sense in this case.

Reply:
Code analysis would protect against that case (Equals returns an ignored value).
I'll lgtm, and we can see how much mileage we can get out of this, but I think immutable model will be less surprise prone.

Reply:
/lgtm

Reply:
I think it's worth considering CEL for this purpose, rather than creating a new DSL.

Reply:
We're using JSONPATH, we didn't create a DSL.